### ISSUE TITLE
GraphNode awareness of it's connections

### ISSUE THREAD

**Issue by nezvers** [godotengine/godot-proposals#2140]
### Describe the project you are working on
Tool for color palette generation using Bezier curves
![image](https://user-images.githubusercontent.com/26794315/104854350-d1405300-590e-11eb-99c3-563205bc20dd.png)

### Describe the problem or limitation you are having in your project
I want to leave logic for each node itself and it needs the ability to check what nodes it has at the input.
One way I probably will need to do is make it aware when connecting in GraphEdit, the other way is going through GraphEdit get_connection_list( ) to find itself.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
Give the node method for getting the input connection list and output connection list.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
```
var connections:Dictionary = get_connection_list(){
    input = [
        {port = 0, node = NodeReference1},
        {port = 2, node = NodeReference2},
    ],
    output = [
        {port = 1, node = NodeReference3},
        {port = 4, node = NodeReference4},
    ]
}
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?
It could be done by making a system and updating at GraphNode connection callback

### Is there a reason why this should be core and not an add-on in the asset library?
It would make working with graph nodes much easier if there was built-in logic for fetching connections

**Comment by pycbouh** [comment]
I don't think it is possible to do internally with how the system is setup. Despite being related, `GraphEdit` and `GraphNode`s are not strictly aware of each other. As you can see from, say, connection methods themselves, nodes are refered to by their name in the tree, not by a reference. Whenever `GraphEdit` needs to do something with its nodes, it just iterates over its children and checks the type. And `GraphNode` interacts with the edit via signals. It's a very idiomatic system for Godot and it doesn't allow for cross-references to be stored.

So a user-side solution is probably the best, with you guaranteeing that a node is inside the edit, and getting a reference to it by calling `get_parent()` and then iterating the connection list for the connections this particular node is involved into. It's not that complicated and you can make helper functions for yourself or even intermediate classes with those helper functions built-in to be used in place of a plain `GraphEdit` or `GraphNode`.

That said, a method of `GraphEdit` that retrieves all connections related to a node can be useful and would reduce the boiler plate. Something like `get_node_connection_list(node_name : String) -> Array`. This method is useful for handling `GraphNode`'s remove requests too.

**Comment by nezvers** [comment]
> I don't think it is possible to do internally with how the system is set up. Despite being related, `GraphEdit` and `GraphNode`s are not strictly aware of each other. As you can see from, say, connection methods themselves, nodes are referred to by their name in the tree, not by a reference.
> 
That also feels like a not good implementation that GraphEdit has only Node.name in the connection list instead of a reference to the node. get_node() call for each entry is stacking up with each connection.
At the moment I have this solution and I don't see why it couldn't be possible to save reference list.
```
# CONNECTION
func _on_GraphEdit_connection_request(from:String, from_slot:int, to:String, to_slot:int)->void:
	connect_node(from, from_slot, to, to_slot)
	var f:GraphNode = get_node(from)
	var t:GraphNode = get_node(to)
	f.outputs[to] = {port = from_slot, node = t}
	t.inputs[from] = {port = to_slot, node = f}
	print(t.inputs.size())

func _on_GraphEdit_disconnection_request(from:String, from_slot:int, to:String, to_slot:int)->void:
	disconnect_node(from, from_slot, to, to_slot)
	var f:GraphNode = get_node(from)
	var t:GraphNode = get_node(to)
	f.outputs.erase(to)
	t.inputs.erase(from)
```
